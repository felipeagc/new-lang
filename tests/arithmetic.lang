def extern vararg printf(_: *u8);

def export main(): i32 {
	// Integer addition

	printf("%d\n",  i32(u8(1 + 2)));
	printf("%d\n", i32(u16(1 + 2)));
	printf("%d\n", i32(u32(1 + 2)));
	printf("%d\n", i32(u64(1 + 2)));

	printf("\n");

	printf("%d\n",  i32(i8(1 + 2)));
	printf("%d\n", i32(i16(1 + 2)));
	printf("%d\n", i32(i32(1 + 2)));
	printf("%d\n", i32(i64(1 + 2)));

	printf("\n");

	// Integer subtraction

	printf("%d\n",  i32(u8(9 - 2)));
	printf("%d\n", i32(u16(9 - 2)));
	printf("%d\n", i32(u32(9 - 2)));
	printf("%d\n", i32(u64(9 - 2)));

	printf("\n");

	printf("%d\n",  i32(i8(9 - 2)));
	printf("%d\n", i32(i16(9 - 2)));
	printf("%d\n", i32(i32(9 - 2)));
	printf("%d\n", i32(i64(9 - 2)));

	printf("\n");
	
	// Integer multiplication

	printf("%d\n",  i32(u8(2 * 3)));
	printf("%d\n", i32(u16(2 * 3)));
	printf("%d\n", i32(u32(2 * 3)));
	printf("%d\n", i32(u64(2 * 3)));

	printf("\n");

	printf("%d\n",  i32(i8(2 * 3)));
	printf("%d\n", i32(i16(2 * 3)));
	printf("%d\n", i32(i32(2 * 3)));
	printf("%d\n", i32(i64(2 * 3)));

	printf("\n");

	// Integer division

	printf("%d\n",  i32(u8(4 / 2)));
	printf("%d\n", i32(u16(4 / 2)));
	printf("%d\n", i32(u32(4 / 2)));
	printf("%d\n", i32(u64(4 / 2)));

	printf("\n");

	printf("%d\n",  i32(i8(4 / 2)));
	printf("%d\n", i32(i16(4 / 2)));
	printf("%d\n", i32(i32(4 / 2)));
	printf("%d\n", i32(i64(4 / 2)));
	
	printf("\n");

	// Integer remainder

	printf("%d\n",  i32(u8(17 % 11)));
	printf("%d\n", i32(u16(17 % 11)));
	printf("%d\n", i32(u32(17 % 11)));
	printf("%d\n", i32(u64(17 % 11)));

	printf("\n");

	printf("%d\n",  i32(i8(17 % 11)));
	printf("%d\n", i32(i16(17 % 11)));
	printf("%d\n", i32(i32(17 % 11)));
	printf("%d\n", i32(i64(17 % 11)));

	return 0;
}
