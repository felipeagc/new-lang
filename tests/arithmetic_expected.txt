3
3
3
3

3
3
3
3

7
7
7
7

7
7
7
7

6
6
6
6

6
6
6
6

2
2
2
2

2
2
2
2

6
6
6
6

6
6
6
6

7.00

3.00

10.00

2.50
